app/src/store/__tests__/audioDevices.test.ts

```diff
@@ -44,73 +56,352 @@ jest.mock('@twilio/voice-react-native-sdk', () => {
 
 describe('audioDevices store', () => {
   let store: Store;
+  let dispatchedActions: any[] = [];
 
   beforeEach(() => {
-    store = createStore();
-  });
+    audioDeviceMap.clear();
+
+    const logAction: Middleware = () => (next) => (action) => {
+      dispatchedActions.push(action);
+      next(action);
+    };
 
-  it('gets an audio device', async () => {
-    await store.dispatch(audioDevices.getAudioDevices());
-    expect(store.getState().voice.audioDevices?.status).toEqual('fulfilled');
+    store = createStore(logAction);
+
+    dispatchedActions.splice(0);
+    jest.clearAllMocks();
   });
 
-  it('handles audio device fetch error', async () => {
-    jest.spyOn(voice, 'getAudioDevices').mockRejectedValue('foo');
-    await store.dispatch(audioDevices.getAudioDevices());
-    expect(store.getState().voice.audioDevices).toEqual({
-      reason: 'GET_AUDIO_DEVICES_ERROR',
-      status: 'rejected',
-      error: miniSerializeError('foo'),
+  const matchDispatchedActions = (actions: any[], actionCreators: any[]) => {
+    expect(actionCreators.map((ac) => ac.type)).toStrictEqual(
+      actions.map((a) => a.type),
+    );
+  };
+
+  describe('updateAudioDevices', () => {
+    it('gets an audio device', async () => {
+      const mockAudioDevices = [
+        { uuid: '1111', type: 'speaker', name: 'device1' },
+        { uuid: '2222', type: 'bluetooth', name: 'device2' },
+        { uuid: '3333', type: 'earpiece', name: 'device3' },
+      ];
+
+      await store.dispatch(
+        audioDevices.updateAudioDevices({
+          audioDevices: mockAudioDevices as any,
+          selectedDevice: mockAudioDevices[1] as any,
+        }),
+      );
+
+      expect(store.getState().voice.audioDevices).toEqual({
+        status: 'fulfilled',
+        audioDevices: mockAudioDevices,
+        selectedDevice: mockAudioDevices[1],
+      });
+      matchDispatchedActions(dispatchedActions, [
+        audioDevices.updateAudioDevices.pending,
+        audioDevices.updateAudioDevices.fulfilled,
+      ]);
+    });
+
+    it('handles no selected device', async () => {
+      const mockAudioDevices = [
```

Comment:
should we move this up one scope so you don't have to keep defining this for every case?

Reply:
We could, but we also modify the devices slightly with every test-case.